Split Presentation thread from Render thread

jroweboy opened the pull request:
Fixes https://github.com/citra-emu/citra/issues/4811 (and maybe https://github.com/citra-emu/citra/issues/4319 https://github.com/citra-emu/citra/issues/3879) superscedes PR https://github.com/citra-emu/citra/pull/4737 as well

Based off the design in https://redream.io/posts/improving-audio-video-synchronization-multi-sync and comments in #4780 this brings shared context support to citra by decoupling guest and host vsync. This does **NOT** include changes to how the framelimiter works, its just changing the vsync situation. Future work can be done to improve audio sync to eliminate popping as well.

### With this every one will now always have vsync on

Previously, citra's framelimiter relied on host vsync being *disabled* in order to work properly, but this is finicky at best. For people who have force enabled vsync in their drivers, citra wasn't actually able to disable vsync, causing inconsistent frame limiting. After this change, the frontend presentation is split on a different thread than the core, and it will draw the latest texture from the mailbox.

### How does the mail box work

in simple terms, we always render the latest frame thats available when its time for host vsync.
if the frontend is faster than the renderer:
  display the previous frame
if the renderer is faster than the frontend:
  drop the queued frame, and replace it with our new one.

Since this changes a lot about how presentation works, I expect issues for various operating systems and GPU vendors, so we shouldn't merge until we can test the following combinations. (Or at least as many as we can)

- [ ] Windows Nvidia
- [ ] Windows AMD
- [ ] Windows Intel
- [ ] macOS Nvidia
- [ ] macOS AMD
- [ ] macOS Intel
- [ ] Linux Nvidia Blob
- [ ] Linux Nvidia Nouveau
- [ ] Linux AMD Blob
- [ ] Linux AMD Mesa
- [ ] Linux Intel Mesa

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/citra-emu/citra/4940)
<!-- Reviewable:end -->

wwylele commented on src/citra_qt/bootmanager.h:
```diff
@@ -17,16 +16,30 @@
 class GMainWindow;
 class GRenderWindow;
 
+class GGLContext : public Frontend::GraphicsContext {
```
nit: why is this named as "GGL"? If this is from **G**RenderWindow... well that is something we wanted to remove (see the very old TODO list)

wwylele commented on src/video_core/renderer_opengl/renderer_opengl.cpp:
```diff
@@ -669,6 +765,38 @@ void RendererOpenGL::DrawScreens(const Layout::FramebufferLayout& layout) {
+
+    /* insert fence for the main thread to block on */
+    frame.present_sync = glFenceSync(GL_SYNC_GPU_COMMANDS_COMPLETE, 0);
+    glFlush();
```
What is this for?

wwylele replied:
got the answer from discord. For future reference: https://www.khronos.org/opengl/wiki/Sync_Object#Flushing_and_contexts

jroweboy replied:
Yeah to quote the relevant part here to save a click: 

> It is important that syncs are properly flushed into the GPU's command queue. Without proper flushing, the sync object may never be signaled... Flushing is ultimately a context-specific action, as it flushes the current context's command queue. Therefore, not only must you perform a flush, you must perform that flush on the same context that created the sync object.

NiceGlassesDude commented:
Tested on: Canary Build - 998b5fa

System: Windows 10, 32 GB ram, gtx 1080

Whenever I switched into fullscreen (F11) the game got into Slowmotion. Everything in the game including audio was extremely slow. Once I switched to windowed mode again, it run pretty fast.

Thanks to @jroweboy I switched the build to 1475 and everything runs smoothly now in fullscreen. It runs much better without stuttering or any kind of lags.

wwylele commented:
Tested on: Canary 1477 vs 1473
System: Windows 10, 16 GB RAM, i7-8750H, GTX 1060
Game tested: Kirby Planet robobot, Monster Hunter XX
Resolution tested: 1x and 4x, windowed or full screen
Summary: no significant performance difference found. If anything, 1477 is slightly better in Kirby (~10% less frame time)

Tenchiro commented:
**System:** windows 7 64bit, i7 4780mq, 8gb ram, HD8790M
**Version tested:** canary-1477

This has performance impact on games ( Games i tested; MHXX, MH4U, MH Gen, legend of zelda Ocarina of time)

**Before canary-1477**
![111](https://user-images.githubusercontent.com/53919721/65062837-75596900-d96c-11e9-9fc0-141b714b8acb.jpg)
![2222](https://user-images.githubusercontent.com/53919721/65062895-9326ce00-d96c-11e9-987c-c2115ccb4b04.jpg)

**Canary 1477**
![3333](https://user-images.githubusercontent.com/53919721/65063464-d46bad80-d96d-11e9-97e3-f788e428ddc6.jpg)
![4444](https://user-images.githubusercontent.com/53919721/65063491-e3526000-d96d-11e9-8842-b75be53dc0ad.jpg)

ivanmestre commented:
Tested on: Canary Build - 998b5fa
System: Windows 10 64bit, i5 6300HQ, 8GB ram, GTX 960M

After this feature was merged performance dropped to around 20% speed on both fullscreen and windowed, was getting full speed before.

![Annotation 2019-09-17 155923](https://user-images.githubusercontent.com/6786286/65071151-6ff82080-d964-11e9-9c2b-52dde9162e79.png)

BreadFish64 commented:
Versions tested: nightly 1358 | canary 1477
System: Windows 10, Intel i7 9750h, 32GB RAM, Nvidia RTX 2060
Performance improves by ~9% in windowed mode, but drops by ~16% in fullscreen.
https://docs.google.com/spreadsheets/d/1dKP8S1ZC6f86KjGVdOjndFE-9AivL6zBf_B7AhwtKKM/edit?usp=sharing

LOuroboros commented:
Canary 1475 and Canary 1477
CPU: i3 2120
RAM: 4 GB
GPU: Gigabyte's NVIDIA GeForce GT 1030 2GB GDDR5
OS: Windows 8.1 Professional x64
Game Tested: Shin Megami Tensei IV v1.1
Internal Resolution: 2x
Areas: Naraku 3rd Stratum, Menus and Battles

- Results -
1475: SMT IV runs at 30 FPS with a Frame Pacing of 11-12ms in both, Windowed and Fullscreen Mode.
1477:  SMT IV runs at 25 to 30 FPS with a Frame Pacing of 17-18ms in both, Windowed and Fullscreen Mode.

jroweboy commented:
Pushed a few macOS fixes, citra should now detect high DPI monitors properly again, and scale accordingly (fixes issue where the game was showing in a tiny corner)

Also I couldn't replicate the slow down but I think I know whats going on and tried to push a fix for that as well. Everyone please retest and check performance in canary 1478 to see if it fixes your issue.

wwylele commented on src/citra_qt/bootmanager.cpp:
```diff
@@ -104,6 +104,10 @@ void GRenderWindow::DoneCurrent() {
 
 void GRenderWindow::PollEvents() {}
 
+void OnFrameSwapped() {
```
leftover code?

LOuroboros commented:
> Everyone please retest and check performance in canary 1478 to see if it fixes your issue.

No notable changes on my end. The frame pace got slightly worse.
25 to 30 with a pace of 18-19ms in Canary 1478.
![citra-qt-2019-09-18_19-53-21](https://user-images.githubusercontent.com/4485172/65192606-081e0480-da4e-11e9-806b-ff8ca1a80510.png)

ivanmestre commented:
Same performance on 1478. Around 20% speed.

jroweboy commented:
Hmm. I spent a good bit of time tonight rewriting the mailbox code, and nothing improved on my laptop. So until I'm ready to test this again, I'm going to remove this from canary so that I can fix any outstanding issues and then put it back on canary when I'm confident that I've fixed the issue.

weihuoya commented:
how about threaded opengl? there will be a solid speed boost.
ppsspp's complex implement, really fast.
https://github.com/hrydgard/ppsspp/pull/10411

GLideN64's general implement, easy to use.
https://github.com/gonetz/GLideN64/pull/2014

wwylele commented:
@weihuoya that has been working on a separate branch

jroweboy commented:
@ Everyone who's tested: I've redo the way the queue works which so far seems to have fixed performance for many people. Please retest and let me know the results.

The freeze on close bug might be fixed too since i rewrote the sync code entirely.

Theres an unintended consequence of these changes which is there is a new soft limit on FPS at 800% speed. I figure thats probably fast enough for most people. But this comes at the cost of creating more textures, which means it eats up additional RAM/VRAM, which is of course multiplied by the size of the window you are rendering into. If we find this takes too much ram, then we can reduce the soft cap to something else. In short, new code makes a speed limit of 800% the max. Increasing that number increases total ram usage based on your citra window size, and likewise decreasing the max speed% will decrease ram usage. So watch your RAM/vRAM usage. (Caveat: read below about resizing)

@LOuroboros If you were resizing the screen, this causes citra to regenerate a ton of textures and buffers, which in turn eats more vRAM/RAM. The old RAM is cleaned up of course, but we can't directly *delete* the old RAM, we essentially tell the driver to delete it and its up to them to reclaim it whenever it feels like. So if you resize the screen, RAM goes up, and it won't go down until the driver feels like doing that.

@weihuoya This isn't related to splitting gpu emulation onto another thread. That will come later if everything plays out well.

wwylele commented on src/video_core/renderer_opengl/gl_state.h:
```diff
@@ -144,6 +144,8 @@ class OpenGLState {
 
     std::array<bool, 2> clip_distance; // GL_CLIP_DISTANCE
 
+    GLuint renderbuffer;
```
```suggestion
    GLuint renderbuffer; // GL_RENDERBUFFER_BINDING
```
(Explanation of this kind of comment in this file: assuming we don't have gl_state but still want to achieve the same effect of reducing opengl call when the state isn't changed, we would use `glGet(xxx)` to check the current state and skip changing the state when the value matches. Of course this is dumb because `glGet` itself is super slow. The comment is the "xxx" if we were to call `glGet`)

wwylele replied:
Also, this needs a corresponding `ResetRenderbuffer` function for resetting the state when the buffer is deleted.

wwylele commented on src/core/frontend/emu_window.h:
```diff
@@ -12,6 +12,61 @@
+    /**
+     * Recreate the presentation objects attached to this frame with the new specified width/height
+     */
+    virtual void ReloadPresentFrame(Frontend::Frame* frame, u32 height, u32 width) = 0;
```
suddenly the height and the width are swapped

wwylele commented on src/video_core/renderer_opengl/renderer_opengl.cpp:
```diff
@@ -28,8 +32,151 @@
+    std::condition_variable free_cv;
+    std::condition_variable present_cv;
+    std::array<Frontend::Frame, SWAP_CHAIN_SIZE> swap_chain{};
+    std::deque<Frontend::Frame*> free_queue{};
```
use `std::queue` to indicate this is a FIFO queue

wwylele commented on src/video_core/renderer_opengl/renderer_opengl.cpp:
```diff
@@ -28,8 +32,151 @@
+        // remove all old entries from the present queue and move them back to the free_queue
+        for (auto f : present_queue) {
+            free_queue.push_back(f);
+            free_cv.notify_one();
```
The notify_one can be moved out of the loop.

wwylele commented on src/video_core/renderer_opengl/renderer_opengl.cpp:
```diff
@@ -28,8 +32,151 @@
+                            [&] { return !present_queue.empty(); });
+        if (present_queue.empty()) {
+            // timed out waiting for a frame to draw so return nullptr
+            return previous_frame;
```
the comment and the code doesn't match.

ivanmestre commented:
@jroweboy Tested latest canary, 1483, slowdown is gone. Back to 100% speed. Great job!!!

mspass commented:
monster hunter gen
canary 1483 FPS 25~ 770MB native res
nightly 1363 FPS 30~ 700MB native res

jroweboy commented:
@mspass please include your hardware (OS CPU and GPU) so i can try to replicate this

mspass commented:
AMD A6-9220 4GB

Dragios commented:
Also sort of solves #4421.

The animation no longer fluctuates with this PR. Although it still did not result in smooth rendering. Animation becomes smooth if you restart the emulation. Which means that if Citra is closed and open back the same game it will result in jittery animation until you restart emulation.

LOuroboros commented:
I'm trying out the build obtained from AppVeyor.
The slowdowns were seemingly fixed. Right now, as I stand inside Alraune's Domain, Citra is using 30% of my CPU and 500-600 MBs of RAM, and the game works just as fine as it used to back in Canary 1475.

jroweboy commented:
@LOuroboros thanks for the testing :+1: its currently on canary again fyi so you can check it out there. It sounds like @Dragios found one issue with it thats going to take a semi-major change to fix, but in theory it should only increase your FPS, so after I change that, I'll probably beg for another retest :D

vvanelslande commented:
Citra version: [Canary 1484](https://github.com/citra-emu/citra-canary/releases/tag/canary-1484) / [AppVeyor 1.0.11166 MinGW](https://ci.appveyor.com/project/bunnei/citra/builds/27565208/job/xkqp74jyx80i25wh/artifacts)
There's no minimum window size.
[![Video](https://cdn-b-east.streamable.com/image/3lrao_first.jpg)](https://streamable.com/3lrao)

jroweboy commented:
Major changes:
Rewrote the qt render target to use QWindow + QOpenGLContext instead of relying on QOpenGLWidget. This might affect performance, and hopefully its for the better. **People with performance issues should retest to make sure everything is okay.** This also includes @Dragios who's issue turned out to be the renderwindow's FPS is lower than his computers vsync, so it caused it to skip frames.

Minor fixes:
Pushed a fix for the minimum size issue.
Addressed review comments (thanks wwylele!)

Last note: 
I'm seriously running out of ideas on how to fix the issues with AMD laptop performance. I'm going to keep trying different ideas but I've only got two more ideas of what could be wrong. It could be poor performance from using shared contexts in general (unlikely, but not fixable), or it could just be that on those cards glBlitFramebuffer is slower than drawing a textured quad (also unlikely, but fixable). But at some point, if we can't fix it, I don't think we should block this PR on this regression.

mspass commented:
same results on my end, in fact i noticed a rather big performance decrease when upping the resolution:
before this PR: 
native: ~30fps - 2x: ~30fps - 3x: ~ 27fps

latest canary:
native: ~25fps - 2x: ~22fps - 3x: ~ 19fps

ivanmestre commented:
Weird bug on latest canary 1485, wasn't doing it on 1483. When going full screen it the first time it works fine, but every single time after it only shows the image cropped with the size of when in Windowed mode.

Windowed mode:
![image](https://user-images.githubusercontent.com/6786286/65388886-73a1f380-dd26-11e9-914e-6c48f2a93eb8.png)

Full screen:
![image](https://user-images.githubusercontent.com/6786286/65388891-86b4c380-dd26-11e9-97d7-1e21d4cb2baa.png)

jroweboy commented:
@mspass can you use an external program that measures FPS? During game play it should show that its matched with your monitors refresh rate which is likely going to be 60FPS. Someone who also has an AMD laptop is saying that its always locked to half of their refresh rate which is really weird, and I want to confirm if this is happening to your machine too.

mspass commented:
so, i use 75hz, which yielded 25fps and 50fps on RivaTuner Statistics Server, so i changed to 60hz, relaunched citra and the framerate locked to 20fps (fluctuating above that) and twice that on RivaTuner.

mspass commented:
just tested usum and got interesting results. on native resolution rivatuner shows 60fps and citra 30fps, on 2x res rivatuner shows 30fps and citra also 30fps.

what i missing here is that the frametime is mostly low, ~20ms on monster hunter gen even at 20fps/60hz and ~14ms on usum (either native or 2x).

Dragios commented:
To whoever that is complaining there is performance issue especially with AMD laptop users, when you guys do the testing, are you on plug in or battery powered? Just wanna make sure power limit is not the issue here.

Because some laptops, (I know majority of modern laptops do) will lower down Host FPS or even half if it detects your laptop runs on battery powered. The reason this happens is to prolong battery life.

Miksel12 commented:
How does this pr affect vrr (variable refresh rate) displays?  If it increase latency, maybe there could be an option to disable any form of host vsync so the vrr monitor can do its thing.

Pik-a-boo commented:
I also have a laptop that uses an AMD graphics card.
I tested Ocarina of Time 3D and it was 30 fps mostly with 2x upscale (some shader stutter here and there).
That was on Windows , I could test on Linux later and edit this post.

Chibicon commented:
Citry Canary 1485 seems to have some issues with the touchscreen when using the screen layout Large Screen. Where you touch on the window isn't accurate. Here's an image. The little pencil icon should be where my click is but it is off some so you have to click below where you actually want to touch.
![JvnSCQ8uSmKmiPijOQswQg 1](https://user-images.githubusercontent.com/1420739/65472221-15653580-de38-11e9-96ef-fd73a531f0f4.png)

jroweboy commented:
Issues with the touch screen should be resolved. Sarcastic shoutouts to this Qt bug https://bugreports.qt.io/browse/QTBUG-68221

I also inserted a glFinish() after swapBuffers is called on the presentation thread. lets see if this has any performance impact (positive or negative) especially for the people reporting slowdowns.

mspass commented:
that definitely improved performance, its almost on par with the latest nightly build usually ~1 frame behind on monster hunter gen, this difference is less apparent on usum and SM3DL.

jroweboy commented:
@mspass which build were you testing? I don't think this build went into the latest canary since there was a conflict

mspass commented:
this latest commit 5b3de00

vvanelslande commented on src/citra_qt/bootmanager.cpp:
```diff
@@ -116,6 +119,8 @@ bool OpenGLWindow::event(QEvent* event) {
     case QEvent::MouseButtonRelease:
     case QEvent::MouseButtonDblClick:
     case QEvent::MouseMove:
+    case QEvent::KeyPress:
```
`QEvent::DragEnter`, `QEvent::DragMove` and `QEvent::Drop` are missing.

Dragios commented:
https://github.com/citra-emu/citra/pull/4940#issuecomment-533811380
An update based on Canary-1487

After the rewrote, Host FPS performance is much better but it still cannot maintain solid 75 FPS (my monitor refresh rate currently) for Pokémon Super Mystery Dungeon game intro. I have to open some program that that can affect CPU multithread usage only then the game can maintain smooth 75 FPS. I'm not sure how accurate is my statement, it could be wrong but this is what I observed.

ivanmestre commented:
> Weird bug on latest canary 1485, wasn't happening on 1483. When going full screen it the first time it works fine, but every single time after it only shows the image cropped with the size of when in Windowed mode.
> 
> Windowed mode:
> ![image](https://user-images.githubusercontent.com/6786286/65388886-73a1f380-dd26-11e9-914e-6c48f2a93eb8.png)
> 
> Full screen:
> ![image](https://user-images.githubusercontent.com/6786286/65388891-86b4c380-dd26-11e9-97d7-1e21d4cb2baa.png)

This is still happening on latest Canary 1487. It also causes the RivaTuner Statistics Server Overlay to show in the middle-left part of the screen instead of the top-left part of the screen.

Might be caused by this 28926cd commit or this 1c45d6c commit.

Pik-a-boo commented:
I got lower FPS after updating to Canary 1487 - MK7 gets 45-50 fps now and before it was 60 fps with 1485 (with upscaling set to Auto).

Developer-exe commented:
Tested on: _Canary 1492_ vs _Nightly 1372_
System: Windows 10, 16 GB RAM, i7-7700HQ, GTX 1050
Game tested: Project X Zone 2
Resolution tested: 1x, windowed or fullscreen
Summary: performance drop in the overworld by as much as 33% (was 60FPS, now hovering around 40-55, never hitting 60FPS); intro and cutscenes elsewhere are relatively stable (50-60FPS)

vitor-k commented on src/citra_qt/main.cpp:
```diff
@@ -2023,11 +2023,21 @@ int main(int argc, char* argv[]) {
     chdir(bin_path.c_str());
 #endif
-
+    QCoreApplication::setAttribute(Qt::AA_EnableHighDpiScaling);
```
This is the reason for #4977  
Was this introduced because of problems with the QOpenGLWidget? Is it still needed?
If it isn't needed anymore, I think it should be removed, as addressing High DPI monitor scaling would probably need a few more changes that would warrant a separate PR .

bunnei commented:
@jroweboy please fix conflicts when you get a chance. Thanks!

jroweboy commented:
@vvanelslande Fixed drag and drop issues
@ivanmestre Fixed full screen issues
@Developer-exe I'm unable to replicate any slow down on any of my three machines. :( Really don't know what to do about that.

LOuroboros commented:
@Developer-exe 
> Game tested: Project X Zone 2
> Resolution tested: 1x, windowed or fullscreen
> Summary: performance drop in the overworld by as much as 33% (was 60FPS, now hovering around 40-55, never hitting 60FPS)

I don't know about 60 FPS because the game seems to run at 30, but I played all the way up to the end of Chapter 1 on Fullscreen at x2 Internal Resolution, and speed-wise the game has worked correctly on my end so far using the latest mingw build for this PR obtained from AppVeyor.

The only "drops" I get are situations where the framerate goes down to 25-28 FPS only for it to go back to 30 right away. The only thing I can tell is that it doesn't seem to be caused by loading assets for the first time, because it happened just now when a character portrait _(Majima's)_ swapped to another character portrait _(Phoenix Wright's)_ even though they both have been showing up on screen for a while now.

Putting that very small issue that I don't even know if this particular PR can even address aside, things have been working flawlessly.
![citra-qt-2019-11-29_11-57-23](https://user-images.githubusercontent.com/4485172/69876875-7c232700-129f-11ea-82ad-051cfd925ca2.png)

Perhaps an uncalled for test, but I figured I should still post the results I got here regardless, in case it helps somehow.
My specs haven't changed. I'm using the same machine in which I tested Shin Megami Tensei IV with this PR back on September.

Developer-exe commented:
> I don't know about 60 FPS because the game seems to run at 30

Ah, my bad, for some reason, my GPU shows twice the actual framerate that's shown by Citra.
![doubleFrameRateShown](https://user-images.githubusercontent.com/14371091/69912953-5d549a00-1431-11ea-8c5c-389e1f6d5346.png)

> I played all the way up to the end of Chapter 1 on Fullscreen at x2 Internal Resolution, and speed-wise the game has worked correctly on my end so far using the latest mingw build for this PR obtained from AppVeyor.

_Canary 1506_ - GPU: 60 FPS, Citra: 30 FPS
[_This PR_](https://ci.appveyor.com/project/bunnei/citra/builds/29197094/job/5xartynenu73utvg/artifacts): GPU: 44-60 FPS, Citra: 25-30 FPS (occasional audio stuttering occurs; framerate often drops under 30 FPS even when left idle during gameplay)

I thought the issue was happening because of the new render thread since it introduced high-DPI scaling issue as well (image below).
![doubleFrameRateShownBigBars](https://user-images.githubusercontent.com/14371091/69913092-a3aaf880-1433-11ea-814e-7d56e726e257.png)
I've always grabbed the latest canary and tested it there. Even after finding this PR on AppVeyor, both issues (incorrect high-DPI scaling and performance drops) are still present.

jroweboy commented:
@Developer-exe As i mentioned, I have 3 machines (AMD Intel and Nvidia) that I've tested these changes on and haven't been able to replicate the slowdown on any of them. I'd love to be able to fix these slowdowns, but I'd first need to be able to replicate it locally. 

We could try a few things if you are available sometime, but I can't guarantee that I'll be able to resolve this issue. In this last commit, I added two new microprofile scopes so that we can try to debug this a little more.

jroweboy commented:
I've rebased #4923 on top of this branch so that we can test this one while we still try to figure out perf regressions with this one. So that means, I'm untagging it here, but it should still be on canary since its part of #4923 (at least until this gets merged)

Developer-exe commented:
> I'd love to be able to fix these slowdowns, but I'd first need to be able to replicate it locally.
> We could try a few things if you are available sometime, but I can't guarantee that I'll be able to resolve this issue.

Sure, I'd love to help with replicating the issue. I should have free Saturdays (living in UTC+1 timezone).

Developer-exe commented:
> Summary: performance drop in the overworld by as much as 33%; intro and cutscenes elsewhere are relatively stable

Here's a [video](https://www.youtube.com/watch?v=QB4l4GxDvmQ) showcasing the issue.

jroweboy commented:
@Developer-exe I haven't changed anything since your last test. I only added some new microprofile scopes to help narrow down where the slowdown is happening. To test this out. Compare nightly versus canary with the following options on

View -> debugging -> Microprofile. then turn on Groups - OpenGL, Timers - Exclusive Time, Exclusive Average

Test the same spot and upload screenshots from microprofile to show me where the slowdown is happening.

DankiestCitra commented:
When running a game on Fullscreen, pressing F5 to stop the emulation causes the screen to freeze the moment the button is pressed. 
The emulation does indeed stop, and Citra Canary continues to run, but the screen is frozen, as if the program isnt successfully displaying whats supposed to display.
All of this happens in Single Window mode.

I should add, that alt-tabbing actually gets the issue fixed, as in, you can see the familiar game selection screen that you have, and start another emulation.

Developer-exe commented:
@jroweboy Here's what I got:

_Canary 1513_:
![CitraMicroProfileCanary1513](https://user-images.githubusercontent.com/14371091/70388505-fc8a0c00-19b2-11ea-81d6-10b18228cf0f.png)

_Nightly 1390_:
![CitraMicroProfileNightly1390](https://user-images.githubusercontent.com/14371091/70388520-41ae3e00-19b3-11ea-8aec-81ad5ff067cc.png)

Developer-exe commented:
@jroweboy There are also some things I've noticed comparing two Citra versions:

Maximum speed:
_Canary 1515_ - 329%
_Nightly 1392_ - 162%

Stable speed:
_Canary 1515_ - 85-95%
_Nightly 1392_ - 95-100%

When the speed limit is set to 400%, the game actually performs better than 100%, but when locked to 100%, _Canary 1515_ performs worse and can't keep up, while _Nightly 1392_ appears to maintain the framerate.

The lower is the speed limit, the less actual speed I get. For example (_Canary 1515_):
400% - stable 329%
300% - stable 184%
200% - stable 181%
100% - Citra reports stable 100%, GPU detects stuttering (can be both seen & heard)

jroweboy commented:
@Developer-exe this sounds very suspicious. Well, heres what I'll do. In the next update, I added a check box back into the UI to turn off VSync. I had another person who had the same issue as you test it, and it seems like thats sufficient to get your speed back. You might get some tearing but I'm out of ideas for whats wrong at this point and why your driver hates vsync so much.

jroweboy commented:
@DankiestCitra fixed your issue as well.
